### PR TITLE
[Feature] Support NDJson and CSV

### LINKED ISSUE #102: NDJSON/CSV methods to add and update documents
⚠️ This issue is generated, it means the nameing might be done differently in this package (ex: `add_documents_json` instead of `addDocumentsJson`). Keep the already existing way of naming in this package to stay idiomatic with the language and this repository.

📣 We strongly recommend doing multiple PRs to solve all the points of this issue

[MeiliSearch v0.23.0](https://github.com/meilisearch/MeiliSearch/releases/tag/v0.23.1) introduces two changes:
- new valid formats to push data files, additionally to the JSON format: CSV and NDJSON formats.
- it enforces the `Content-type` header for every route requiring a payload (`POST` and `PUT` routes)

---

Here are the expected changes to completely close the issue:

- [ ] Currently, the SDKs always send `Content-Type: application/json` to every request. Only the `POST` and `PUT` requests should send the `Content-Type: application/json` and not the `DELETE` and `GET` ones.

- [ ] Add the following methods and 🔥 **the associated tests** 🔥 to ADD the documents. Depending on the format type (`csv` or `ndjson`) the SDK should send `Content-Type: application/x-dnjson` or `Content-Type: text/csv`)
  - [ ] `addDocumentsJson(string docs, string primaryKey)`
  - [ ] `addDocumentsCsv(string docs, string primaryKey)`
  - [ ] `addDocumentsCsvInBatches(string docs, int batchSize, string primaryKey)`
  - [ ] `addDocumentsNdjson(string docs, string primaryKey)`
  - [ ] `addDocumentsNdjsonInBatches(string docs, int batchSize, string primaryKey)`

- [ ] Add the following methods and 🔥 **the associated tests** 🔥 to UPDATE the documents. Depending on the format type (`csv` or `ndjson`) the SDK should send `Content-Type: application/x-dnjson` or `Content-Type: text/csv`)
  - [ ] `updateDocumentsJson(string docs, string primaryKey)`
  - [ ] `updateDocumentsCsv(string docs, string primaryKey)`
  - [ ] `updateDocumentsCsvInBatches(string docs, int batchSize, string primaryKey)`
  - [ ] `updateDocumentsNdjson(string docs, string primaryKey)`
  - [ ] `updateDocumentsNdjsonInBatches(string docs, int batchSize, string primaryKey)`

`docs` are the documents sent as `String`
`primaryKey` is [the primary key of the index](https://docs.meilisearch.com/learn/core_concepts/indexes.html#primary-key)
`batchSize` is the size of the batch. Example: you can send 2000 documents in raw String in `docs` and ask for a `batchSize` of 1000, so your documents will be sent to MeiliSearch in two batches.

Example of PRs:
- in PHP SDK: https://github.com/meilisearch/meilisearch-php/pull/235
- in Python SDK: https://github.com/meilisearch/meilisearch-python/pull/329

----------

Related to: https://github.com/meilisearch/integration-guides/issues/146

If this issue is partially/completely implemented, feel free to let us know.

### PR DESCRIPTION
**Pull request by ahmednfwela**
Fixes #102 
Depends on #261 
- [x] Tests Added

### CODE REVIEW

File: lib/src/index_impl.dart [hunk 1]
```diff
@@ -158,43 +157,217 @@ class MeiliSearchIndexImpl implements MeiliSearchIndex {
...
+  @override
+  Future<Task> addDocumentsJson(String documents, {String? primaryKey}) {
+    final decoded = jsonDecode(documents);
+    if (decoded is List<Object?>) {
+      final casted = decoded.whereType<Map<String, Object?>>().toList();
+      return addDocuments(casted, primaryKey: primaryKey);
+    }
+    throw MeiliSearchApiException(
+      "Provided json must be an array of documents, consider using addDocumentsNdjson if this isn't the case",
+    );
```
Comment (brunoocasali) [inline lib/src/index_impl.dart:179]:
```suggestion
    final decoded = jsonDecode(documents);

    if (decoded is List<Object?>) {
      final casted = decoded.whereType<Map<String, Object?>>().toList();

      return addDocuments(casted, primaryKey: primaryKey);
    }

    throw MeiliSearchApiException(
      "Provided json must be an array of documents. Consider using addDocumentsNdjson if this isn't the case",
    );
```

File: lib/src/index_impl.dart [hunk 2]
```diff
@@ -158,43 +157,217 @@ class MeiliSearchIndexImpl implements MeiliSearchIndex {
...
+
+  @override
+  Future<List<Task>> addDocumentsNdjsonInBatches(
+    String documents, {
+    String? primaryKey,
+    int batchSize = 1000,
+  }) {
+    final ls = LineSplitter();
+    final split = ls.convert(documents);
+    return Future.wait(
```
Comment (brunoocasali) [inline lib/src/index_impl.dart:247]:
```suggestion

    return Future.wait(
```

File: lib/src/index_impl.dart [hunk 3]
```diff
@@ -158,43 +157,217 @@ class MeiliSearchIndexImpl implements MeiliSearchIndex {
...
+
+  @override
+  Future<List<Task>> addDocumentsCsvInBatches(
+    String documents, {
+    String? primaryKey,
+    int batchSize = 1000,
+  }) {
+    final ls = LineSplitter();
+    final split = ls.convert(documents);
+    return Future.wait(
```
Comment (brunoocasali) [inline lib/src/index_impl.dart:229]:
```suggestion

    return Future.wait(
```

File: lib/src/index_impl.dart [hunk 4]
```diff
@@ -158,43 +157,217 @@ class MeiliSearchIndexImpl implements MeiliSearchIndex {
...
+    String? primaryKey,
+  }) {
+    final decoded = jsonDecode(documents);
+    if (decoded is List<Object?>) {
+      final casted = decoded.whereType<Map<String, Object?>>().toList();
+      return updateDocuments(casted, primaryKey: primaryKey);
+    }
+    throw MeiliSearchApiException(
+      "Provided json must be an array of documents, consider using updateDocumentsNdjson if this isn't the case",
+    );
```
Comment (brunoocasali) [inline lib/src/index_impl.dart:283]:
```suggestion
    final decoded = jsonDecode(documents);

    if (decoded is List<Object?>) {
      final casted = decoded.whereType<Map<String, Object?>>().toList();

      return updateDocuments(casted, primaryKey: primaryKey);
    }

    throw MeiliSearchApiException(
      "Provided json must be an array of documents. Consider using updateDocumentsNdjson if this isn't the case",
    );
```

File: lib/src/index_impl.dart [hunk 5]
```diff
@@ -158,43 +157,217 @@ class MeiliSearchIndexImpl implements MeiliSearchIndex {
...
+
+  @override
+  Future<List<Task>> updateDocumentsNdjsonInBatches(
+    String documents, {
+    String? primaryKey,
+    int batchSize = 1000,
+  }) {
+    final ls = LineSplitter();
+    final split = ls.convert(documents);
+    return Future.wait(
```
Comment (brunoocasali) [inline lib/src/index_impl.dart:348]:
```suggestion

    return Future.wait(
```

File: test/documents_test.dart [hunk 6]
```diff
@@ -37,6 +52,31 @@ void main() {
       expect(doc?['title'], equals('The Hobbit 2'));
     });
 
+    test('Update documents in batches', () async {
+      const batchSize = 10;
```
Comment (brunoocasali) [inline test/documents_test.dart:56]: Can you apply that xUnit concept here as well?

**Review by brunoocasali — COMMENTED** [review]
I left some comments, but in fact I think solving the conflicts will help a lot!

File: lib/src/index.dart [hunk 7]
```diff
@@ -54,19 +54,130 @@ abstract class MeiliSearchIndex {
...
     List<Map<String, Object?>> documents, {
     String? primaryKey,
   });
 
+  /// Add a list of documents by given [documents] and optional [primaryKey] parameter
+  ///
+  /// * the passed [documents] must be a valid JSON string representing an array of objects.
+  /// * If index does not exist, it tries to create a new index and adds documents.
+  Future<Task> addDocumentsJson(
+    String documents, {
```
[the remark this answers is not in the document]
Comment (ahmednfwela) [reply lib/src/index.dart:71]: The current `addDocuments` already receives a parsed json, all these new methods mainly target file contents, so all of them must be raw strings.
or is that not the case ?

**Review by brunoocasali — APPROVED** [review]
Thanks a lot @ahmednfwela 🎉 

bors merge

**Pull request merged**